File: motor3d 2.0.ConvertedToC#/Vertice.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Xml.Linq;
using Motor3D.Espacio2D;
using Motor3D.Espacio3D;
using Motor3D.Espacio3D.Transformaciones;
using Motor3D.Escena;
using Motor3D.Escena.Shading;
using System.Drawing;

namespace Motor3D.Primitivas3D
{
	public class Vertice : ObjetoPrimitiva3D
	{

		private Punto3D mCoordenadasSUR;
		private Punto3D mCoordenadasSRC;
		private Punto2D mRepresentacion;
		private Vector3D mNormalSUR;
		private Vector3D mNormalSRC;
		private Color mColor;
		private Color mColorShading;
		private int[] mCaras;

		private bool mShaded;
		public Punto3D CoodenadasSUR {
			get { return mCoordenadasSUR; }
			set {
				mCoordenadasSUR = value;
				mShaded = false;
			}
		}

		public Punto3D CoodenadasSRC {
			get { return mCoordenadasSRC; }
			set {
				mCoordenadasSRC = value;
				mShaded = false;
			}
		}

		public Punto2D Representacion {
			get { return mRepresentacion; }
			set { mRepresentacion = value; }
		}

		public Vector3D NormalSUR {
			get { return mNormalSUR; }
			set {
				mNormalSUR = value;
				mShaded = false;
			}
		}

		public Vector3D NormalSRC {
			get { return mNormalSRC; }
			set {
				mNormalSRC = value;
				mShaded = false;
			}
		}

		public Color Color {
			get { return mColor; }
			set {
				mColor = value;
				mShaded = false;
			}
		}

		public Color ColorShading {
			get { return mColor; }
			set {
				mColorShading = value;
				mShaded = true;
			}
		}

		public int[] Caras {
			get { return mCaras; }
		}

		public int Caras {
			get {
				if (Indice >= 0 && Indice <= mCaras.GetUpperBound(0)) {
					return mCaras[Indice];
				} else {
					return -1;
				}
			}
		}

		public bool Shaded {
			get { return mShaded; }
		}

		public void AplicarTransformacion(Transformacion3D Transformacion)
		{
			mCoordenadasSUR *= Transformacion;
			mShaded = false;
		}

		public void EstablecerCaras(params int[] Caras)
		{
			mCaras = Caras;
		}

		public void Shading(PhongShader Constantes, Foco3D[] Focos, Camara3D Camara)
		{
			mColorShading = Constantes.EcuacionPhong(Focos, mNormalSRC, mCoordenadasSRC, mColor, Camara);
		}

		public Vertice(Punto3D Coordenadas)
		{
			mCoordenadasSUR = Coordenadas;
			mRepresentacion = new Punto2D();
			mNormalSUR = new Vector3D();
			mColor = Color.White;
			mColorShading = Color.White;
			mShaded = false;
		}

		public static Punto3D BaricentroSUR(params Vertice[] Vertices)
		{
			double x = 0;
			double y = 0;
			double z = 0;

			x = 0;
			y = 0;
			z = 0;

			for (int i = 0; i <= Vertices.GetUpperBound(0); i++) {
				x += Vertices[i].CoodenadasSUR.X;
				y += Vertices[i].CoodenadasSUR.Y;
				z += Vertices[i].CoodenadasSUR.Z;
			}

			return new Punto3D(x / (Vertices.GetUpperBound(0) + 1), y / (Vertices.GetUpperBound(0) + 1), z / (Vertices.GetUpperBound(0) + 1));
		}

		public static Punto3D BaricentroSRC(params Vertice[] Vertices)
		{
			double x = 0;
			double y = 0;
			double z = 0;

			x = 0;
			y = 0;
			z = 0;

			for (int i = 0; i <= Vertices.GetUpperBound(0); i++) {
				x += Vertices[i].CoodenadasSRC.X;
				y += Vertices[i].CoodenadasSRC.Y;
				z += Vertices[i].CoodenadasSRC.Z;
			}

			return new Punto3D(x / (Vertices.GetUpperBound(0) + 1), y / (Vertices.GetUpperBound(0) + 1), z / (Vertices.GetUpperBound(0) + 1));
		}

		public override string ToString()
		{
			return "{Vertice de Coordenadas=" + mCoordenadasSUR.ToString() + ", Normal=" + mNormalSUR.ToString() + " y Representacion=" + mRepresentacion.ToString() + "}";
		}
	}
}

File: motor3d 2.0/vertice.vb
Imports Motor3D.Espacio2D
Imports Motor3D.Espacio3D
Imports Motor3D.Espacio3D.Transformaciones
Imports Motor3D.Escena
Imports Motor3D.Escena.Shading
Imports System.Drawing

Namespace Primitivas3D
    Public Class Vertice
        Inherits ObjetoPrimitiva3D

        Private mCoordenadasSUR As Punto3D
        Private mCoordenadasSRC As Punto3D
        Private mRepresentacion As Punto2D
        Private mNormalSUR As Vector3D
        Private mNormalSRC As Vector3D
        Private mColor As Color
        Private mColorShading As Color
        Private mCaras() As Integer
        Private mShaded As Boolean

        Public Property CoodenadasSUR() As Punto3D
            Get
                Return mCoordenadasSUR
            End Get
            Set(ByVal value As Punto3D)
                mCoordenadasSUR = value
                mShaded = False
            End Set
        End Property

        Public Property CoodenadasSRC() As Punto3D
            Get
                Return mCoordenadasSRC
            End Get
            Set(ByVal value As Punto3D)
                mCoordenadasSRC = value
                mShaded = False
            End Set
        End Property

        Public Property Representacion() As Punto2D
            Get
                Return mRepresentacion
            End Get
            Set(ByVal value As Punto2D)
                mRepresentacion = value
            End Set
        End Property

        Public Property NormalSUR() As Vector3D
            Get
                Return mNormalSUR
            End Get
            Set(ByVal value As Vector3D)
                mNormalSUR = value
                mShaded = False
            End Set
        End Property

        Public Property NormalSRC() As Vector3D
            Get
                Return mNormalSRC
            End Get
            Set(ByVal value As Vector3D)
                mNormalSRC = value
                mShaded = False
            End Set
        End Property

        Public Property Color() As Color
            Get
                Return mColor
            End Get
            Set(ByVal value As Color)
                mColor = value
                mShaded = False
            End Set
        End Property

        Public Property ColorShading() As Color
            Get
                Return mColor
            End Get
            Set(ByVal value As Color)
                mColorShading = value
                mShaded = True
            End Set
        End Property

        Public ReadOnly Property Caras() As Integer()
            Get
                Return mCaras
            End Get
        End Property

        Public ReadOnly Property Caras(ByVal Indice As Integer) As Integer
            Get
                If Indice >= 0 AndAlso Indice <= mCaras.GetUpperBound(0) Then
                    Return mCaras(Indice)
                Else
                    Return -1
                End If
            End Get
        End Property

        Public ReadOnly Property Shaded() As Boolean
            Get
                Return mShaded
            End Get
        End Property

        Public Sub AplicarTransformacion(ByVal Transformacion As Transformacion3D)
            mCoordenadasSUR *= Transformacion
            mShaded = False
        End Sub

        Public Sub EstablecerCaras(ByVal ParamArray Caras() As Integer)
            mCaras = Caras
        End Sub

        Public Sub Shading(ByVal Constantes As PhongShader, ByVal Focos() As Foco3D, ByVal Camara As Camara3D)
            mColorShading = Constantes.EcuacionPhong(Focos, mNormalSRC, mCoordenadasSRC, mColor, Camara)
        End Sub

        Public Sub New(ByVal Coordenadas As Punto3D)
            mCoordenadasSUR = Coordenadas
            mRepresentacion = New Punto2D
            mNormalSUR = New Vector3D
            mColor = Color.White
            mColorShading = Color.White
            mShaded = False
        End Sub

        Public Shared Function BaricentroSUR(ByVal ParamArray Vertices() As Vertice) As Punto3D
            Dim x, y, z As Double

            x = 0
            y = 0
            z = 0

            For i As Integer = 0 To Vertices.GetUpperBound(0)
                x += Vertices(i).CoodenadasSUR.X
                y += Vertices(i).CoodenadasSUR.Y
                z += Vertices(i).CoodenadasSUR.Z
            Next

            Return New Punto3D(x / (Vertices.GetUpperBound(0) + 1), y / (Vertices.GetUpperBound(0) + 1), z / (Vertices.GetUpperBound(0) + 1))
        End Function

        Public Shared Function BaricentroSRC(ByVal ParamArray Vertices() As Vertice) As Punto3D
            Dim x, y, z As Double

            x = 0
            y = 0
            z = 0

            For i As Integer = 0 To Vertices.GetUpperBound(0)
                x += Vertices(i).CoodenadasSRC.X
                y += Vertices(i).CoodenadasSRC.Y
                z += Vertices(i).CoodenadasSRC.Z
            Next

            Return New Punto3D(x / (Vertices.GetUpperBound(0) + 1), y / (Vertices.GetUpperBound(0) + 1), z / (Vertices.GetUpperBound(0) + 1))
        End Function

        Public Overrides Function ToString() As String
            Return "{Vertice de Coordenadas=" & mCoordenadasSUR.ToString & ", Normal=" & mNormalSUR.ToString & " y Representacion=" & mRepresentacion.ToString & "}"
        End Function
    End Class
End Namespace

